draft-ietf-tls-esni.md

```diff
@@ -306,6 +310,118 @@ servers to rotate the keys often and improve forward secrecy.
+each possibly of size 1, that match the IPv4 or IPv6 address of a host with the private key.
+They are encoded using the following structure.
+
+~~~~
+    enum {
+        address_v4(4),
+        address_v6(6),
+    } AddressType;
+
+    struct {
+        AddressType address_type;
+        select (address_type) {
+            case address_v4: {
+                uint8 ipv4Length;
+                opaque ipv4Address[4];
+            }
+            case address_v6: {
+                uint8 ipv6Length;
+                opaque ipv6Address[16];
+            }
+        }
+    } Address;
+
+    struct {
+        HostName name;
+        Address address_set<0..2^16-1>;
+    } HostPointer;
+~~~~
+
+name
+: The terminal name for which resolution must yield valid A or
+AAAA records pointing to a host which holds the private ESNI key.
+
+address_set
+: An optional list of Address entries, each containing a single IPv4 or IPv6
+CIDR block, possibly of length 1.
+
+Use of this structure during the ESNI resolution algorithm is described in {{esni-resolution}}.
+
+### Resolution Algorithm
+
+Clients obtain ESNI records by querying the DNS for ESNI-enabled server domains.
+In cases where the domain of the A or AAAA records being resolved do not match the
+SNI Server Name, such as when {{!RFC7838}} is being used, the alternate domain should
+be used for querying the ESNI TXT record.
+
+Clients SHOULD initiate ESNI queries in parallel alongside normal A or AAAA queries.
+The following algorithm describes a procedure by which clients can process ESNIKeys
+responses as they arrive to produce addresses for ESNI-capable hosts.
+
+~~~
+1. If an ESNIKeys response arrives before an A or AAAA response, and the ESNIKeys
+response contains a "host_pointer" extension with a full address, clients SHOULD 
+initiate TLS with ESNI to the provided address(es).
+
+2. If an A or AAAA response arrives before the ESNIKeys response, clients SHOULD
+wait up to CD seconds before initiating TLS to either address. (Clients may begin
+TCP connections in this time. QUIC connections should wait.) If an ESNIKeys
+response does not arrive in this time, clients SHOULD initiate TLS without ESNI
+to the provided address(es).
+
+3. If an ESNIKeys response with a "host_pointer" extension arrives before this time, 
+and if the address netmask matches that in the corresponding A or AAAA response, 
+then clients SHOULD initiate TLS with ESNI to said address(es). (An ESNIKeys record 
+with a full address has a full netmask, and therefore the match must be exact.) 
+Clients MAY delay this check until both A and AAAA responses arrive, or if a 
+Resolution Delay timeout occurs, according to {{RFC8305}}.
+
+4. If the address netmask does not match the corresponding A or AAAA response,
+and the A or AAAA response yielded a terminal CNAME that matches host_pointer.name,
+then clients SHOULD initiate TLS using ESNI to the address(es) in the A or AAAA records.
```

Comment:
So the idea here is that, even if there's no overlap between the A/AAAA results and the ESNI record addresses, if the host_pointer chain includes (terminal might be too strong a statement) the indicated name, then you try it?  This is a bit of a leap of faith on two fronts:  First, that they have the appropriate keys without being in the list of addresses, but second that the caller will get the CNAME chain back rather than just the addresses.

Perhaps a reasonable optimization for those clients that can do it, though.